tests/unit/s2n_send_test.c

```diff
@@ -53,21 +55,144 @@ static int s2n_track_sent_bytes_partial_send_fn(void *io_context, const uint8_t
         return -1;
     }
 
+    s2n_custom_send_fn_called = true;
+
     int partial_read = len-3;
 
     sent_bytes = partial_read;
     errno = EAGAIN;
 
+    return partial_read;
+}
+
+static int s2n_fail_send_with_injected_errno_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    errno = *(int*)io_context;
+    s2n_custom_send_fn_called = true;
+
+    return S2N_FAILURE;
+}
+
+static int s2n_send_alert_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->in));
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->alert_in));
+    EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->in, buf, len));
+
+    /* We expect a fatal alert to be sent. */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_process_alert_fragment(conn), S2N_ERR_ALERT);
+    EXPECT_TRUE(conn->closed);
+
+    return len;
+}
+
+static int s2n_send_always_passes_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->out));
+    EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->out, buf, len));
+
+    return len;
+}
+
+static int s2n_send_mitigates_beast_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+    uint32_t *writes = (uint32_t*) s2n_connection_get_ctx(conn);
+
+    /* The BEAST mitigation in s2n_send should first send a 1 byte record and then the rest of the application data. */
+    uint32_t expected_write_sizes[] = {1, 12};
+
+    EXPECT_EQUAL(conn->current_user_data_consumed, expected_write_sizes[*writes]);
+
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->out));
+    EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->out, buf, len));
+
+    *writes += 1;
+
+    return len;
+}
+
+static int s2n_partial_socket_send_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    uint32_t *written_bytes = (uint32_t*) s2n_connection_get_ctx(conn);
+
+    /* Break loop on second call. */
+    if (s2n_custom_send_fn_called) {
+        errno = EAGAIN;
+        return -1;
+    }
+
     s2n_custom_send_fn_called = true;
+    int partial_read = len-3;
+
+    *written_bytes = partial_read;
+    errno = EAGAIN;
 
     return partial_read;
 }
 
+static int s2n_dynamic_record_sizing_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+    uint32_t *writes = (uint32_t*) s2n_connection_get_ctx(conn);
+    
+    uint16_t min_payload_size = 0;
+    POSIX_GUARD_RESULT(s2n_record_min_write_payload_size(conn, &min_payload_size));
+    EXPECT_EQUAL(1398, min_payload_size);
```

Comment:
I'll add a comment to say as much but this is here because this number is the max amount of bytes that can be stuffed into a single ethernet frame. 

So when dynamic mode is used it will use this as the max write size until the bytes written threshold is reached.

I have this check here so if for some reason this value changes the test case can be revisited (Will add a comment for this as well)